### PR TITLE
Game optmisations

### PR DESCRIPTION
**Pull request by brenjohn**
This PR is looking to speed up the select move methods of the mcts and brandubh_zero bots in order to speed up training and evaluation.

The main optimisation made was to translate the game submodule into cython and refactor the GameState and GameSet methods for moving game pieces and computing legal moves for a given game state. Some small refactoring was also made to the brandubh_zero methods for scoring branches.

These changes resulted in a reasonable speed up in the mcts select move method while that of the zero bot is roughly equivalent. (Note, a mistake was also corrected in the legal_moves function which made both select_move methods a good bit slower. So while the timings below show no real speed up for the zero bot, there is technically a speed up since these methods are now doing more work. Namely, before, when legal moves were being computed, the logic for removing pieces when a piece was moved wasn't being called. This has been fixed.)

select_move_time_vs_evals

Some other changes which might makes sense to implement going forward might be:
- refactoring the GameState class to replace history linked list with a dictionary whose fast lookup might provide a more efficient way to check if a given board state has happened already. This would have to be tested to see if the extra copying of history dictionaries doesn't cancel any performance benefits offered by the fast look up.
- Some extra pruning could be implemented in bot mcts and zero tree algorithms for choosing nodes to explore. Namely, when a terminal state is found, all other child nodes stemming from the parent node should be removed.
- Profiling shows that, apart from the tensorflow invocations, the methods for scoring branches in the zero algorithm are the most expensive, so optimising these (maybe with cython) might improve things a bit.
- One of the most expensive methods for the mcts algorithm is the greedy bot method for finding winning moves from a given list of legal moves. It would be easy to implement a version of the method for computing legal moves which also returned which legal moves were winning moves. This would not only significantly speedup the greedy and mcts bots, but also might help implement the above pruning idea efficiently in both mcts and zero bots.

### CODE REVIEW

**Comment by brenjohn** [comment]
The legal_moves method of the GameState class has been refactored to return both a list of legal moves and a list of winning moves. The GreeadyRandom, MCTS and BrandubhZero bots have also been refactored to take advantage of this feature. Namely, the GreedyBot now selects moves much faster and both the MCTS and BrandubhZero bots use the feature to prune branches when winning moves exist.

Additionally, some futrher optimisation to reduce python overhead in GameSet and GameState methods have been made and the HistoryLink class was refactored to optimise scanning the game history (most importantly the state representation was changed from the bytes of an int array of length 49 to an int[3] variable - the bits of the three integers encoding the position of indistinguishable game pieces)

These changes have produced a significant speed up in the time needed to select a move for each bot as shown below.

select_move_time_vs_evals

The history linked list was not replaced by a dictionary as suggested above as the time spent of searching the game history is negligible.

The branch scoring methods have also not been modified as previously suggested as the time spent of scoring branches is small compared to the time spent on neural network calls.
